### PR TITLE
chore: Add MISRA-2012 check using cppcheck's misra addon.

### PR DESCRIPTION
**Pull request by iphydf**
Disabled a whole bunch of rules from the MISRA-C set. Some of them
should be fixed, but most of the ones we violate have good reasons. This
PR documents those reasons.

---
This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2251)

### CODE REVIEW

File: other/docker/misra/Makefile [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+#
+# Reason: This warns about ((unsigned)n == 0) and other constant comparisons.
+SUPPRESSIONS += 10.4
+# The value of a composite expression shall not be cast to a different essential type category or a wider essential type.
+#
+# TODO(iphydf): investigate.
+SUPPRESSIONS += 10.8
+# A conversion should not be performed from pointer to void into pointer to object.
+#
+# Reason: this is needed for generic callbacks to make any sense.
```
Comment (sudden6) [inline other/docker/misra/Makefile:43]: maybe we should keep our internal API free from casts to void? AFAICT this does not apply to objects passed in externally and then stored as `void *userdata`

File: other/docker/misra/Makefile [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+# TODO(iphydf): Why is this a good idea?
+SUPPRESSIONS += 15.7
+# An unconditional break statement shall terminate every switch-clause.
+#
+# Reason: This conflicts with "break unused after abort()". MISRA doesn't allow
+# abort(), but we use it, so this rule must be disabled, too.
+SUPPRESSIONS += 16.3
+# Every switch statement shall have a default label.
+#
+# TODO(iphydf): maybe fix?
```
Comment (sudden6) [inline other/docker/misra/Makefile:89]: IIRC default may conflict with a diagnostic from compilers that check that alll enum values are covered. I don't think that should be fixed by MISRA rules, but rather by using that compiler diagnostic.

File: other/docker/misra/Makefile (hunk 1 above)
Comment (iphydf) [reply other/docker/misra/Makefile:43]: Lots of our internal APIs use internal callbacks. E.g. look for all the `(DHT *)` casts. It's everywhere. I'm not sure it's realistic to stop that, but maybe we can make a linter that says only objects named "object", "userdata", or "user_data" can be cast-from-void, to restrict the usage a bit?

File: other/docker/misra/Makefile (hunk 2 above)
Comment (iphydf) [reply other/docker/misra/Makefile]: Yeah, I agree. Removed the TODO.

File: other/docker/misra/Makefile (hunk 1 above)
Comment (sudden6) [reply other/docker/misra/Makefile:43]: A linter for only `userdata` sounds like a good compromise. I'm not meaning to change this in the short term anyway.

**Review by sudden6 — APPROVED** [review]

**Pull request merged**